### PR TITLE
Add target info to "Library.change!" pubsub message

### PR DESCRIPTION
**Pull request by byewokko**
# Summary
- `LibraryService.subscribe` is extended with optional `target` argument.
- Possible `target` values:
  - `None` (default) or `"global"` to watch global target changes,
  - `"tenant"` to watch changes in ANY tenant,
  - `("tenant", TENANT_ID)` to watch changes in a specific tenant.
- The callback method must expect `target` argument, whose value is either `"global"` or `("tenant", TENANT_ID)`.
- Backward compatibility: When  `LibraryService.subscribe` is called without `target` argument (or its value is `None`), the callback is called without `target` argument.

# Other changes
- Fix: Superuser has access to any tenant.

### CODE REVIEW

File: asab/library/providers/zookeeper.py [hunk 1]
```diff
@@ -340,15 +348,53 @@ def recursive_traversal(path, digest):
...
+					path = "/.tenants/{}{}".format(tenant, path)
+				try:
+					newdigest = await self._get_directory_hash(path)
+				except kazoo.exceptions.NoNodeError:
+					# This node is either deleted or has never existed.
+					newdigest = None
+
+				if newdigest != digests.get(tenant):
+					self.Subscriptions[path][tenant] = newdigest
+					ctx = Tenant.set(tenant)
```
Comment (ateska) [inline asab/library/providers/zookeeper.py:369]: This will not work for sync->async PubSub

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/library/providers/zookeeper.py (hunk 1 above)
Comment (byewokko) [reply asab/library/providers/zookeeper.py:369]: context will be replaced with `target` argument

tenant targets:
```python
self.App.PubSub.publish("Library.change!", self, path, target=("tenant", tenant))
```

global targets:
```python
self.App.PubSub.publish("Library.change!", self, path)  # For backward compatibility
```

File: asab/library/providers/zookeeper.py [hunk 2]
```diff
@@ -156,7 +157,7 @@ def __init__(self, library, path, layer):
 		# This is a contingency check for changes for subscribed folders in library even without version counter change.
 		self.App.PubSub.subscribe("Application.tick/600!", self._on_library_changed)
 
-		self.Subscriptions = {}
+		self.Subscriptions: typing.Dict[str, typing.Dict[str, bytes]] = {}
```
Comment (byewokko) [inline asab/library/providers/zookeeper.py:160]: no nested dicts, use tuple as key instead:

```python
self.Subscriptions = {
  ("global", "/Templates/Email"): b"abc123",
  ("tenant", "/Reports/Whatevs", "my-tenant"): b"def456",
  ("tenant", "/Reports/Whatevs", "your-tenant"): b"ghi789",
  ...
}
```

File: asab/library/providers/zookeeper.py [hunk 3]
```diff
@@ -314,7 +315,14 @@ def build_path(self, path):
 
 
 	async def subscribe(self, path):
```
Comment (byewokko) [inline asab/library/providers/zookeeper.py:317]: add optional argument `target` with possible values:
- `"global"` (default) to watch global target changes
- `"tenant"` to watch ANY tenant changes
- `("tenant", TENANT_ID)` to watch a specific tenant changes

File: asab/library/providers/zookeeper.py (hunk 2 above)
Comment (byewokko) [reply asab/library/providers/zookeeper.py]: replaced with a set of subscriptions and a dict of node hashes

File: asab/library/providers/zookeeper.py
Comment (language-shprt) [inline asab/library/providers/zookeeper.py:1]: The same information is repeated 2 times, lines 101 - 112: https://github.com/TeskaLabs/asab/pull/603/files#diff-fb1b9ac09cf3f76afce8a8a0afa4d8c19895cd3a626d80692fcb2d337e992c70R101-R112

Comment (language-shprt) [reply asab/library/providers/zookeeper.py:1]: Also, the whole Example is probably obsolete? We don't do it as `tenant_token = asab.library.providers.zookeeper.TenantContextVar.set` anymore

Comment (byewokko) [reply asab/library/providers/zookeeper.py:1]: true, i'll update it with the new context var

**Review by language-shprt — APPROVED** [review]

**Pull request merged**